oxidized-transformers/src/models/hf/checkpoint.rs

```diff
@@ -54,95 +62,209 @@ where
+            HFCheckpoint::Multiple(PYTORCH_INDEX.into())
+        } else {
+            HFCheckpoint::Single(PYTORCH_SINGLE.into())
+        };
+
+        let paths = checkpoint.load(self)?;
+
+        Ok(Box::new(
+            PyTorchTensors::multi(&paths).context(LoadCheckpointSnafu)?,
+        ))
+    }
+}
+struct PyTorchTensors {
+    tensors: Vec<PthTensors>,
+    routing: HashMap<String, usize>,
+}
 
-    /// Load sharded safetensors checkpoint.
-    fn load_safetensors_multi(&self, index_path: &Path) -> Result<Vec<PathBuf>, BoxedError> {
-        // Parse the shard index.
-        let index_file = BufReader::new(
-            File::open(index_path).context(OpenSafeTensorsIndexSnafu { path: index_path })?,
-        );
-        let index: SafeTensorsIndex = serde_json::from_reader(index_file)
-            .context(ParseSafeTensorsIndexSnafu { path: index_path })?;
-
-        let shard_names = index.shards();
-        let mut shards = Vec::with_capacity(shard_names.len());
-        for shard_name in shard_names {
-            let path = self.file(&shard_name).context(DownloadSnafu {
-                name: shard_name.clone(),
-            })?;
-            ensure!(path.is_some(), NonExistentShardSnafu { name: shard_name });
-            shards.push(path.unwrap());
+impl PyTorchTensors {
+    fn multi(paths: &[impl AsRef<Path>]) -> Result<Self, candle_core::Error> {
+        let mut routing = HashMap::new();
+        let mut tensors = vec![];
+
+        for (index, p) in paths.iter().enumerate() {
+            // We need to read the tensor metadata separately as PthTensors
+            // does not expose this information.
+            let tensor_names: Vec<String> = read_pth_tensor_info(p, false, None)?
+                .into_iter()
+                .map(|ti| ti.name.clone())
+                .collect();
+
+            let pth = PthTensors::new(p, None)?;
+            for k in tensor_names {
+                routing.insert(k, index);
+            }
+            tensors.push(pth);
         }
+        Ok(Self { tensors, routing })
+    }
 
-        Ok(shards)
+    fn load(&self, name: &str, dev: &Device) -> Result<Tensor, candle_core::Error> {
+        self.get(name)?.to_device(dev)
     }
 
-    /// Load non-sharded safetensors checkpoint.
-    fn load_safetensors_single(&self) -> Result<Vec<PathBuf>, BoxedError> {
-        let path = self.file(SAFETENSORS_SINGLE).context(DownloadSnafu {
-            name: SAFETENSORS_SINGLE,
-        })?;
+    fn get(&self, name: &str) -> Result<Tensor, candle_core::Error> {
+        let index =
+            *self
+                .routing
+                .get(name)
+                .ok_or_else(|| candle_core::Error::CannotFindTensor {
+                    path: name.to_string(),
+                })?;
 
-        ensure!(
-            path.is_some(),
-            NonExistentCheckpointSnafu {
-                name: SAFETENSORS_SINGLE.to_string(),
+        Ok(self.tensors[index].get(name)?.unwrap())
```

Comment:
Is this unwrap safe? I think `None` will only be returned when the name does not exist (which can't occur here, it'd fail in the routing). But upstream code can change, so maybe it's best to be defensive here and also return `CannotFindTensor` here when it's `None`?

Reply:
Fair point, I fix that.